### PR TITLE
Added: Support incoming connection without source node for AWS 

### LINKED ISSUE #5: Support incoming connection without source node
An edge is defined from one Node to an other Node. What if we have incoming connections from the `outside` (SSH from `0.0.0.0/0`, 80/443 connection on LB) ?

We could create a dedicated node for `the outside`. 

This logic should belong to the `provider.Provider.ResourceInOut`

### PR DESCRIPTION
**Pull request by xescugc**
What this PR does it add the ability to show the external connections (via `ingress`) to the nodes.

This is just implemented by AWS for now and this is how it looks right now:

graph

The name of the Nodes is `inframap_outside.p:{PROTOCOL}-f{FROM_PORT}:-t{TO_PORT}-{RESOURCE_CONNECTED_TO}` we should decide:
* If we want it (should be a name unique on the Graph)
* If we want to show **ALL** the incomming connections (as it it now) or just show one to represent that there is an access from the outside and nothing else.

Closes #5

### CODE REVIEW

File: provider/type.go [hunk 1]
```diff
@@ -8,6 +8,7 @@ type Type int
 // List of all the Providers supported
 const (
 	Raw Type = iota
+	InfraMap
```
Comment (xlr-8) [inline provider/type.go:11]: Not sure to understand the addition of this provider?

**Review by xlr-8 — COMMENTED** [review]
LGTM and it's nice! However and if possible - I'd tend to prefer 1 Node with multiple rules.

I believe here you have several rules from an identical "fictional" node, but that can make the reading quite hard to read. I'm not a big fan of the naming either, I'd remove anything that is redundant. Considering this has to have a unique name, including the port/protocol/destination should be enough.

To be fair it would make even more sense to have the port on the connections themselves rather than on the node, but I don't think it's possible with such format. Same goes for connection from front to redis, mysql & so on.

**Comment by xlr-8** [comment]
> So for each security_group in this case it would be represented by 1 Node even if it have multiple ingress no?

For example yeah, 1 node with the naming mentioning those 3 rules and only 1 connection.

> Should be, but it's not (I've tested it and found some cases of redundancy). I don't like the name either I'll try the one you mention again or find a way to be able to define a name without having to be unique on the Graph (this is a rule that we have).

OK

> Yes that would be the best solution, it is possible with Graphviz to define those but right now we do not support it. Your idea would be to have one Node with 2 edges to the aws_elb.front one with 80:80 and the other with the 443:443 no?

For example yeah, or even only one if that can be represented as such, the tricky thing is always about keeping it 'human readable'. So perhaps we could have a question of: when do we decide to split those connection from one into X - X being the limit of readable (2, 3 - we define it)

**Comment by xescugc** [comment]
Ok I've pushed a new version in which I changed some things:

graph

* The names will be `inframap_outside.p:{PROTOCOL}-f{FROM_PORT}:-t{TO_PORT}`
* Each Node is an `ingress` block. And also this ingress block is shared between several other SG, so a Node will potentially have several connections, so instead of having 30 Nodes that are "equal" we'll have 1 Node with 30 edges to it which makes it more readable for big infras.

2 questions:
* Should we change the name of the node? Basically the `inframap_outside` the `outside` if we want to change it to something different.
* Should we add a flag to not do that logic? By default it should be `true` IMO but as the `--connections=false` have a way to now show those. The idea would be `--outside=false` (or whatever the name we decide to use).

**Comment by xescugc** [comment]
At the end the output will be 
inframap

File: cmd/generate.go [hunk 2]
```diff
@@ -95,4 +97,5 @@ func init() {
 	generateCmd.Flags().BoolVar(&clean, "clean", true, "Clean will the generated graph will not have any Node that does not have a connection/edge")
 	generateCmd.Flags().BoolVar(&connections, "connections", true, "Connections will apply the logic of the provider to remove resources that are not nodes")
 	generateCmd.Flags().BoolVar(&showIcons, "show-icons", true, "Toggle the icons on the printed graph")
+	generateCmd.Flags().BoolVar(&externalNodes, "external-nodes", true, "Toggle the addtion of extranal nodes like 'im_out' (used to show ingress connections)")
```
Comment (xlr-8) [inline cmd/generate.go:100]: external*

**Comment by xescugc** [comment]
Ready to review.

File: generate/state.go [hunk 3]
```diff
@@ -372,6 +376,59 @@ func fixEdges(g *graph.Graph, cfg map[string]map[string]interface{}, opt Options
...
+			if e.Target != id {
+				continue
+			}
+			for _, no := range nodes {
+				pv, _, err := getProviderAndResource(no, opt)
+				if err != nil {
+					return err
+				}
+
+				res, err := pv.Resource(no)
```
Comment (xlr-8) [inline generate/state.go:397]: I think the naming could be improved here.

I assumed by reading it that `getProviderAndResource` would already return a `*resource.Resource` as second argument, while it seems to simply be the `Name` of the Resource, which is not what we want - thus the need of calling `pv.Resource` afterwards.

Why not returning the full Resource? Also, why are `IsNode` / `IsEdge` have a Provider as a receiver rather than a `Resource` wouldn't it make more sense?

File: provider/flexibleengine/flexibleengine.go [hunk 4]
```diff
@@ -101,7 +101,7 @@ func (a Provider) ResourceInOut(id, rs string, cfgs map[string]map[string]interf
 			ins = append(ins, pid.(string))
 		}
 	}
-	return ins, outs
+	return ins, outs, nil
```
Comment (xlr-8) [inline provider/flexibleengine/flexibleengine.go:104]: The change of prototype/return is just for AWS isn't it?

File: cmd/generate.go (hunk 2 above)
Comment (xescugc) [reply cmd/generate.go:100]: and also the `addition`.

File: provider/flexibleengine/flexibleengine.go (hunk 4 above)
Comment (xescugc) [reply provider/flexibleengine/flexibleengine.go:104]: Yes for now.

Comment (xescugc) [reply provider/flexibleengine/flexibleengine.go:104]: I did also update the README as I forgot :)

File: generate/state.go (hunk 3 above)
Comment (xescugc) [reply generate/state.go:397]: This are not changes of this PR but ok haha (meaning that it's just used logic not defined).

~First of all there is no such thing as `*resource.Resource` on IM so that would not be an issue.~ Ok you mean a `tfdocs/resource.Resource`. The easy answer is that we do not want the `resource.Resource` in most of the cases.

`getProviderAndResource`: With param `aws_instance.front` would return `aws.Provider, "aws_instance"`

`IsNode/IsEdge`: It's on the provider as it's the one that has the logic (it's on `providers/aws/graph.go`) and the `resource.Rresource` is from TFDocs and it's more related to a Resource documentation.

Basically the `tfdocs/resource.Resource` only has information of the Documentation of the resource, it has a `Metadata map[string]struct{}` that we could push a `"node": true|false` and add a function to that repositry to check that, but I feel it'll create an inderect connection between the 2 libs `tfdcos` and `inframap` which I would try to avoid and keep all the IM logic inside IM.

**Comment by xescugc** [comment]
Done/Commented.

**Comment by tormath1** [comment]
LGTM, you can rebase !

**Review by xlr-8 — APPROVED** [review]

**Comment by xescugc** [comment]
Ok fixed I'm gonna merge then :+1:
